tools/release/adot-operator-images-mirror/mirror.go

```diff
@@ -0,0 +1,244 @@
+
+	// Fetch remote tags from source repository.
+	m.remoteTags, err = m.getRemoteTags()
+	if err != nil {
+		return err
+	}
+
+	// Filter the tags we got.
+	m.filterTags()
+
+	return nil
+}
+
+func (m *mirror) pullImage(tag string) error {
+	pullOptions := docker.PullImageOptions{
+		Repository:        m.sourceRepositoryFullName(),
+		Tag:               tag,
+		InactivityTimeout: 1 * time.Minute,
+	}
+
+	return (*m.dockerClient).PullImage(pullOptions, docker.AuthConfiguration{})
+}
+
+func (m *mirror) tagImage(tag string) error {
+	tagOptions := docker.TagImageOptions{
+		Repo:  m.targetRepositoryName(),
+		Tag:   tag,
+		Force: true,
+	}
+
+	return (*m.dockerClient).TagImage(m.sourceRepositoryFullName() + ":" + tag, tagOptions)
+}
+
+func (m *mirror) pushImage(tag string) error {
+	pushOptions := docker.PushImageOptions{
+		Name:              m.targetRepositoryName(),
+		Registry:          m.targetRepo.Registry,
+		Tag:               tag,
+		InactivityTimeout: 1 * time.Minute,
+	}
+
+	// Get ecrpublic credentials.
+	creds, err := getDockerCredentials(ecrPublic)
+	if err != nil {
+		return err
+	}
+
+	return (*m.dockerClient).PushImage(pushOptions, *creds)
+}
+
+func (m *mirror) work() {
+	if err := m.ecrManager.ensure(m.targetRepo.Name); err != nil {
+		log.Fatalf("Failed to create ECR repo %s: %v", m.targetRepo.Name, err)
+	}
+
+	for _, tag := range m.remoteTags {
+		if err := m.pullImage(tag.Name); err != nil {
+			log.Printf("Failed to pull docker image %s:%s: %v", m.sourceRepositoryFullName(), tag.Name, err)
+			continue
+		}
+
+		if err := m.tagImage(tag.Name); err != nil {
+			log.Printf("Failed to retag docker image %s:%s: %v", m.sourceRepositoryFullName(), tag.Name, err)
+			continue
+		}
+
+		if err := m.pushImage(tag.Name); err != nil {
+			log.Printf("Failed to push retagged image %s:%s: %v", m.targetRepositoryName(), tag.Name, err)
+			continue
+		}
+	}
+}
+
+func (m *mirror) getRemoteTags() ([]RepositoryTag, error) {
+	var url string
+	switch m.sourceRepo.Host {
+	case quay:
+		url = fmt.Sprintf("https://quay.io/api/v1/repository/%s/tag", m.sourceRepositoryName())
+	case gcr:
+		url = fmt.Sprintf("https://gcr.io/v2/%s/tags/list", m.sourceRepositoryName())
+	}
+
+	var (
+		allTags []RepositoryTag
+		err     error
+		res     *http.Response
+		req     *http.Request
+		retries = 5
+	)
+
+	// Try to fetch the tags information through HTTP get request.
+	for retries > 0 {
+		req, err = http.NewRequest("GET", url, nil)
+		if err != nil {
+				return nil, err
+			}
+
+		res, err = httpClient.Do(req)
+		if err != nil {
+			log.Printf("Failed to get %s, retrying", url)
+			retries--
+		} else if res.StatusCode == 429 {
+			sleepTime := getSleepTime(res.Header.Get("X-RateLimit-Reset"), time.Now())
+			log.Printf("Rate limited on %s, sleeping for %s", url, sleepTime)
+			time.Sleep(sleepTime)
+			retries--
+		} else if res.StatusCode < 200 || res.StatusCode >= 300 {
+			log.Printf("Get %s failed with %d, retrying", url, res.StatusCode)
```

Comment:
We can have some sleep time for all retries.

Reply:
Got it. Let me add this.